DevCycle/Models/CustomData.swift

```diff
@@ -0,0 +1,110 @@
+//  Created by Jonathan Norris on 2022-01-24.
+//  Copyright © 2022 Taplytics. All rights reserved.
+//
+
+import Foundation
+
+public enum CustomDataValue: Codable {
+    case string(_ value: String)
+    case number(_ value: NSNumber)
+    case boolean(_ value: Bool)
+    
+    public init(from decoder: Decoder) throws {
+        let singleValueContainer = try decoder.singleValueContainer()
+
+        do {
+            let value = try singleValueContainer.decode(Double.self)
+            self = .number(NSNumber(value: value))
+            return
+        } catch {}
+        
+        do {
+            let value = try singleValueContainer.decode(Bool.self)
+            self = .boolean(value)
+            return
+        } catch {}
+        
+        do {
+            let value = try singleValueContainer.decode(String.self)
+            self = .string(value)
+            return
+        } catch {}
+        
+        throw DecodingError.dataCorruptedError(
+            in: singleValueContainer,
+            debugDescription: "Invalid CustomDataValue, must be a String / Number / Boolean"
+        )
+    }
+    
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        switch self {
+        case .string(let value):
+            return try container.encode(value)
+        case .number(let value):
+            return try container.encode(Double(value))
+        case .boolean(let value):
+            return try container.encode(value)
+        }
+    }
+}
+
+public class CustomData: Codable {
+    public var data: [String: CustomDataValue]
+    
+    public static func customDataFromDic(_ data: [String: Any]) throws -> CustomData {
+        let data = try JSONSerialization.data(withJSONObject: data)
+        let decoder = JSONDecoder()
+        return try decoder.decode(CustomData.self, from: data)
+    }
+    
+    required public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: DynamicCodingKeys.self)
+        var tmpData = [String: CustomDataValue]()
+        
+        for key in container.allKeys {
+            let decodedObject = try container.decode(
+                CustomDataValue.self,
+                forKey: DynamicCodingKeys(stringValue: key.stringValue)!
+            )
+            tmpData[key.stringValue] = decodedObject
+        }
+        self.data = tmpData
+    }
+    
+//    public init(_ data: [String: Any]) throws {
+//        let data = try JSONSerialization.data(withJSONObject: data)
+//        let decoder = JSONDecoder()
+//        self.data = try decoder.decode(Self.self, from: data)
+//    }
+    
+    private struct DynamicCodingKeys: CodingKey {
+        // Use for string-keyed dictionary
+        var stringValue: String
+        init?(stringValue: String) {
+            self.stringValue = stringValue
+        }
+
+        // Use for integer-keyed dictionary
+        var intValue: Int?
+        init?(intValue: Int) {
```

Comment:
if we don't need this, is this necessary to include the init?

Reply:
let me check.

Reply:
yea its needed: 
![image](https://user-images.githubusercontent.com/1219069/150852091-b1a89572-3721-4813-b339-3798f2245281.png)